### PR TITLE
variables exercise

### LINKED ISSUE #14: Variables
## Brief
Ansible uses variables to enable more flexibility in playbooks and roles. They can be used to loop through a set of given vales, access various information like the hostname of a system and replace certain strings in templates by system-specific values. 

We need to explain the different ways that variables can by used through:
  - the inventory
  - by variable files
  - on the command line
  - retrieved from the vault

## Suggested Exercise
Variables in loops
Variables in templates
Variables in conditions (when)
Getting variables from the system (facts)
Getting variables from the command line
Setting variables in playbooks
Setting variables in the inventory
Setting variables on a system (local facts)
Results of tasks (registered variables)
Accessing variables of other hosts

### Introduction
Give an explanation of the different ways in which variables can be used (see list above in Suggested Exercise)

### Tutorial
Variable types:
 - k/v pairs
 - dictionaries

### Verification test
Define a test that confirms that the participant has correctly used variables in the exercise.

## Other
 Good doc here: https://liquidat.wordpress.com/2016/01/26/howto-introduction-to-ansible-variables/

### PR DESCRIPTION
**Pull request by lashd**
This resolved #14.

I have made some changes @heathtechnical's initial draft:
- Adding additional content
- Reordering a couple of the sections to improve the flow
- Expanding on some of the examples
- Changing acceptance test to cover the exercise's end task

### CODE REVIEW

File: exercises/IaC/ansible/variables/.templates/README.md.erb [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+  - manage differences across systems
+  - keep code and configuration separate.
+  - organise your configuration.
+
+## Useful Terminology
+- DRY (Don't Repeat Yourself) - keeping code DRY is a very simple but important idea, that being that the duplication is best avoided so keep your code DRY and don't repeat yourself.
+
+**Note:** Before starting the exercises, please do the following:
+  - `cd YOUR_CLONE_OF_THIS REPO`
+  - `source ./bin/env`
```
Comment (jimconner) [inline exercises/IaC/ansible/variables/.templates/README.md.erb]: file is currently called .env rather then env - This is just a reminder to fix.

File: exercises/IaC/ansible/variables/.templates/README.md.erb [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+  - `cd <%=env('exercise_path') %>`
+  - `<%= command("cic up") %>`
+
+
+Having run this command, test infrastructure consisting of all the environments you will need has been stood up for you.
```
Comment (jimconner) [inline exercises/IaC/ansible/variables/.templates/README.md.erb]: This sentence parses strangely in my head... Having run 'this' command (but I just ran four commands)... Maybe reword to 'The `cic up` command just provisioned the environment you will need to complete this exercise'

File: exercises/IaC/ansible/variables/.templates/README.md.erb [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+
+Runing <%= after_rendering_run(cic_down_command) %> will stop these environments and drop any changes that you have made to them.
```
Comment (jimconner) [inline exercises/IaC/ansible/variables/.templates/README.md.erb]: As I suggested rewording the cic up text, then we could single-case 'these environments' to 'the environment' instead.

File: exercises/IaC/ansible/variables/.templates/README.md.erb [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+
+```YAML
+<%= hello_world_yaml %>
+```
+
+Against the 'Display a variable' task you'll notice a string containing `{{name_variable}}` is being passed to the `debug` module. This notation denotes a variable called: `name_variable` is expected to be supplied at runtime.
+
+**Note:** Playbooks are processed using the [Jinja2 templating engine](http://jinja.pocoo.org/docs/2.10/). With the exception of the `when:` attribute, variables are always referenced via `{{ VARIABLE }}` notation.
+
+There are number of different ways of supply the value for a variable
```
Comment (jimconner) [inline exercises/IaC/ansible/variables/.templates/README.md.erb]: different way _to_ supply the value

File: exercises/IaC/ansible/variables/.templates/README.md.erb [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+
+### Supplying from variables through the commandline
+Write the above ansible to a file <%= write_to_file(hello_world_path, hello_world_yaml) %> and run: <%= command(%Q{ansible-playbook #{hello_world_path} -c local --extra-vars="name_variable=Micky"} ) %>
```
Comment (jimconner) [inline exercises/IaC/ansible/variables/.templates/README.md.erb]: The 'ansible' folder that the hello_world_path includes does not exist at this stage. Maybe we want to commit a .gitkeep inside it so that people don't have to work out that they need to do a mkdir?

File: exercises/IaC/ansible/variables/.templates/README.md.erb [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+<%= prompt_playbook %>
+```
+
+Here we can see that the `include_vars` module is being used to load the variables. The file that the variables should be loaded from are determined by inspecting the value of the `app_env` variable at run time. The `app_env` variable could be set against each server definition in the inventory for example. Write the following to <%= write_to_file('ansible/inventory', inventory_yaml) %>, which does just that.
+<% %>
+```YAML
+<%= inventory_yaml %>
+```
+
+Before running the playbook, run `cic connect <%= dev_server %>` and note that the prompt does contain the text `<%= dev_prompt %>`. After executing the playbook by running `<%=command("ansible-playbook ansible/prompt.yml -i ansible/inventory")%>`, connect back to <%= dev_server %> and you will see that the prompt now reads as required.
```
Comment (jimconner) [inline exercises/IaC/ansible/variables/.templates/README.md.erb]: Also... note that the prompt does not contain _dev_ .. but as the box is called dev1 then actually it does contain _dev_ ... `root@dev1:/cic#`

Comment (jimconner) [reply exercises/IaC/ansible/variables/.templates/README.md.erb]: Also, prompt after running exercise is uppercase `[DEV]$`

File: exercises/IaC/ansible/variables/.templates/README.md.erb [hunk 7]
```diff
@@ -0,0 +1,378 @@
...
+- prod-app1
+- prod-app2
+
+The application deployed to these environments requires a file to be located at `/etc/app.id`. This file should containing a single line in the format:
+
+```APPSTRING:ENVIRONMENT:HOSTNAME```
+
+- `APPSTRING` is a static string - almost always `ABCAPP`, but sometimes the team need to change this for ad-hoc testing, so they would like this to default to ```CIC_WEBAPP``` but be overridable from the command line.
+
+- `ENVIRONMENT` is the environment that the server is a member of. This should be DEV for development machines and PROD for production machines.
```
Comment (jimconner) [inline exercises/IaC/ansible/variables/.templates/README.md.erb]: _environment_ appears to be a special word in ansible and is not resolved as a normal variable - It is worth having a note to this effect, or perhaps using app_env instead of _environment_ to that people don't go down the wrong path by default.

Comment (jimconner) [reply exercises/IaC/ansible/variables/.templates/README.md.erb]: Should probably call out to the doc about restricted keywords here as _environment_ is not the only one. https://docs.ansible.com/ansible/2.5/reference_appendices/playbooks_keywords.html

File: exercises/IaC/ansible/variables/.templates/README.md.erb (hunk 1 above)
Comment (lashd) [reply exercises/IaC/ansible/variables/.templates/README.md.erb]: I've renamed this in the branch.

File: exercises/IaC/ansible/variables/.templates/README.md.erb (hunk 2 above)
Comment (lashd) [reply exercises/IaC/ansible/variables/.templates/README.md.erb]: I've changed this to:

**Note:** Before starting the exercises, please do the following:
  - `cd YOUR_CLONE_OF_THIS REPO`
  - `source ./bin/env`
  - `cd <%=env('exercise_path') %>`
  - `<%= command("cic up") %>` - This command provisions the test environments that you will need for this exercise.

File: exercises/IaC/ansible/variables/.templates/README.md.erb (hunk 3 above)
Comment (lashd) [reply exercises/IaC/ansible/variables/.templates/README.md.erb]: Agreed

File: exercises/IaC/ansible/variables/.templates/README.md.erb (hunk 4 above)
Comment (lashd) [reply exercises/IaC/ansible/variables/.templates/README.md.erb]: I have reworded this to: There are many different ways of supplying variable in Ansible

File: exercises/IaC/ansible/variables/.templates/README.md.erb [hunk 8]
```diff
@@ -0,0 +1,378 @@
...
+
+## Scope/Precidence
+Ansible has 3 mains scopes:
+- Global - These are values set in config, environment variables or on the commandline
+- Per play
+- Per host
+
+More details on how Ansible handles variable scope can be found in the [Ansible Documentation](https://docs.ansible.com/ansible/latest/user_guide/playbooks_variables.html#variable-scopes).
+
+In Ansible, variables can be declared in a number of different places and have a different precedence depending on where they are defined. For example, variables set via the commandline have the highest precedence. In the given the example above, values for the `name` and `greeting` were defined at the playbook level. These can be overidden using the variables defined at the commandline level which running `<%= command(%Q{ansible-playbook -c local #{hello_world_path} --extra-vars='name_variable=Daffy'}) %>` shows.
```
[the remark this answers is not in the document]
Comment (lashd) [reply exercises/IaC/ansible/variables/.templates/README.md.erb]: The sentence that followed the quoted one instructed readers to write the supplied playbook to file. I have reworded the section 'Supplying variables through the commandline' to make it more explicit

File: exercises/IaC/ansible/variables/.templates/README.md.erb (hunk 6 above)
[the remark this answers is not in the document]
Comment (lashd) [reply exercises/IaC/ansible/variables/.templates/README.md.erb]: fixed

Comment (lashd) [reply exercises/IaC/ansible/variables/.templates/README.md.erb]: I have fixed this so that DEV is written uppercase

File: exercises/IaC/ansible/variables/.templates/README.md.erb (hunk 7 above)
Comment (lashd) [reply exercises/IaC/ansible/variables/.templates/README.md.erb]: I have put in a note warning participants against choosing a variable name that clashes with those used by ansible:

> **Note:** Be careful not to choose a variable name that is one of [the variable names used by Ansible](https://docs.ansible.com/ansible/2.5/reference_appendices/playbooks_keywords.html) as doing so can lead to unexpected results.

File: exercises/IaC/ansible/variables/.templates/README.md.erb [hunk 9]
```diff
@@ -0,0 +1,378 @@
...
+When `<%= cic_up_command %>` was run, the following environments with given hostnames were  stood up:
+
+- <%= dev_server %>
+- prod1
+- prod2
+
+To connect to any of these environment simply use the `<%=cic_connect_command %>` connect command, e.g.  `<%= cic_connect_command %> <%= dev_server %>`
+
+
+Write the following development specific config to `<%= write_to_file('ansible/vars/dev.yml', dev_yaml) %>`
```
[the remark this answers is not in the document]
Comment (lashd) [reply exercises/IaC/ansible/variables/.templates/README.md.erb]: done
